test/infrastructure/vcsim/controllers/controlplaneendpoint_controller.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	inmemoryruntime "sigs.k8s.io/cluster-api-provider-vsphere/test/infrastructure/tmp/runtime"
+	inmemoryserver "sigs.k8s.io/cluster-api-provider-vsphere/test/infrastructure/tmp/server"
+	vcsimv1 "sigs.k8s.io/cluster-api-provider-vsphere/test/infrastructure/vcsim/api/v1alpha1"
+)
+
+type ControlPlaneEndpointReconciler struct {
+	Client client.Client
+
+	InMemoryManager inmemoryruntime.Manager
+	APIServerMux    *inmemoryserver.WorkloadClustersMux
+	PodIP           string
+
+	// WatchFilterValue is the label value used to filter events prior to reconciliation.
+	WatchFilterValue string
+}
+
+// +kubebuilder:rbac:groups=vcsim.infrastructure.cluster.x-k8s.io,resources=controlplaneendpoints,verbs=get;list;watch;patch
+// +kubebuilder:rbac:groups=vcsim.infrastructure.cluster.x-k8s.io,resources=controlplaneendpoints/status,verbs=get;update;patch
+
+func (r *ControlPlaneEndpointReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the ControlPlaneEndpoint instance
+	controlPlaneEndpoint := &vcsimv1.ControlPlaneEndpoint{}
+	if err := r.Client.Get(ctx, req.NamespacedName, controlPlaneEndpoint); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(controlPlaneEndpoint, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Always attempt to Patch the controlPlaneEndpoint object and status after each reconciliation.
+	defer func() {
+		if err := patchHelper.Patch(ctx, controlPlaneEndpoint); err != nil {
+			log.Error(err, "failed to patch ControlPlaneEndpoint")
+			if reterr == nil {
+				reterr = err
+			}
```

Comment:
```suggestion
			reterr = kerrors.NewAggregate([]error{reterr, err})
```

(we recently changed this in ~ all our controllers)

Reply:
Same in other controllers